agent/lib/kontena/workers/log_worker.rb

```diff
@@ -17,70 +19,69 @@ class LogWorker
     ETCD_PREFIX = '/kontena/log_worker/containers'
     QUEUE_MAX_SIZE = 2000
     QUEUE_THROTTLE = (QUEUE_MAX_SIZE * 0.8)
+    WATCH_INTERVAL = 10.0
 
     # @param [Boolean] autostart
     def initialize(autostart = true)
       @queue = Queue.new
       @workers = {}
       @etcd = Etcd.client(host: '127.0.0.1', port: 2379)
+      @running = nil
       subscribe('container:event', :on_container_event)
       subscribe('websocket:connected', :on_connect) # from master_info RPC
       subscribe('websocket:disconnected', :on_disconnect)
       info 'initialized'
 
-      async.start if autostart
+      if autostart
+        async.watch_queue
+        async.start
+      end
+    end
+
+    # @return [Boolean]
+    def running?
+      !!@running
     end
 
     # @param [String] topic
     # @param [Object] data
     def on_connect(topic, data)
-      @queue_processing = true
-      async.start
-      info 'started log streaming'
+      start unless running?
     end
 
     # @param [String] topic
     # @param [Object] data
     def on_disconnect(topic, data)
-      @queue_processing = false
-      async.stop
-      info 'stopped log streaming'
-    end
-
-    def queue_processing?
-      @queue_processing == true
+      stop if running?
     end
 
     def start
-      sleep 1 until Actor[:etcd_launcher].running?
-      Docker::Container.all.each do |container|
-        begin
-          self.stream_container_logs(container) unless container.skip_logs?
-        rescue Docker::Error::NotFoundError
-          # Could be thrown since container.skip_logs? actually loads the container details
-        end
-      end
-
-      async.watch_queue
-      async.process_queue
-    end
+      @running = true
 
-    def watch_queue
-      loop do
-        sleep 10
-        if @queue.size > QUEUE_MAX_SIZE
-          warn "queue is full (size is #{@queue.size}), log lines are dropped until queue has free space"
-        elsif @queue.size > QUEUE_THROTTLE
-          warn "queue size is #{@queue.size}, log streams are throttled and some log lines may be dropped"
-        elsif @queue.size > (QUEUE_MAX_SIZE / 2)
-          warn "queue size is #{@queue.size}"
+      exclusive {
+        info 'start log streaming'
+
+        wait_until!("etcd running") { Actor[:etcd_launcher].running? }
+
+        Docker::Container.all.each do |container|
+          begin
+            self.stream_container_logs(container) unless container.skip_logs?
+          rescue Docker::Error::NotFoundError => exc
+            # Could be thrown since container.skip_logs? actually loads the container details
+            warn exc.message
+          rescue => exc
+            error exc
+          end
         end
-      end
+
+        async.process_queue
+      }
     end
 
+    # Process items from @queue until nil
     def process_queue
       defer {
-        while queue_processing? && data = @queue.pop
+        while data = @queue.pop
```

Comment:
Bad: this change means that it will empty out the `@queue` and drop any queued up log items when disconnecting, because the `@queue << nil` from `stop` will show up after any queued up messages.

Reply:
Not anymore.